File: cpp/questions/kthSmallest.cpp
#include <functional>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int ans;
        std::function<void(TreeNode*)> io = [&](TreeNode* n) -> void {
            if (n == nullptr) {
                return;
            }
            io(n->left);
            if (--k == 0) {
                ans = n->val;
                return;
            }
            io(n->right);
        };
        io(root);
        return ans;
    }
};

File: python/questions/kthSmallest.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        ans = 0
        def io(n: Optional[TreeNode]) -> None:        
            nonlocal k, ans
            if n is None:
                return
            io(n.left)
            k -= 1
            if k == 0:
                ans = n.val
                return
            io(n.right)
        io(root)
        return ans
